<a href="https://colab.research.google.com/github/Amritanshu-Ranjan/Amritanshu-Ranjan/blob/main/Case_Study_Stack_OverFlow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import all important Libraries needed for the Case Study
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
PythonVersion = !python --version

In [2]:
# Validate the Version of Libraries Used
print('          -- VERSION CHECK --             ')
print('Python Version     :' , PythonVersion[0])
print('Pandas Version     :' , pd.__version__)
print('Numpy Version      :' , np.__version__)
print('Seaborn Version    :' , sns.__version__)
print('Matploltib Version :' , matplotlib.__version__)

          -- VERSION CHECK --             
Python Version     : Python 3.7.11
Pandas Version     : 1.1.5
Numpy Version      : 1.19.5
Seaborn Version    : 0.11.1
Matploltib Version : 3.2.2


***The DataSets Provided To Us As Part of Case Study***

1. *Questions.csv*: Different questions asked by the users

  - https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Questions.csv

2. *Answers.csv*: All the answers that are given to the aforementioned questions

  - https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Answers.csv

3. *Tags.csv*: The various tags of the questions that are asked 

  - https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Tags.csv

4. *Users.csv*: All the users that were present between the mentioned duration

  - https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Users.csv

In [3]:
questions_df = pd.read_csv('https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Questions.csv',encoding='ISO-8859-1')
answers_df   = pd.read_csv('https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Answers.csv',encoding='ISO-8859-1')
tags_df      = pd.read_csv('https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Tags.csv',encoding='ISO-8859-1')
users_df     = pd.read_csv('https://stack-overflow-casestudy-datasets.s3.amazonaws.com/Users.csv',encoding='ISO-8859-1')

In [5]:
questions_df.head()

Id  ...                                               Body
0   80  ...  <p>I've written a database generation script i...
1   90  ...  <p>Are there any really good tutorials explain...
2  120  ...  <p>Has anyone got experience creating <strong>...
3  180  ...  <p>This is something I've pseudo-solved many t...
4  260  ...  <p>I have a little game written in C#. It uses...

[5 rows x 7 columns]

In [6]:
answers_df.head()

Id  OwnerUserId  ... Score                                               Body
0   92         61.0  ...    13  <p><a href="http://svnbook.red-bean.com/">Vers...
1  124         26.0  ...    12  <p>I wound up using this. It is a kind of a ha...
2  199         50.0  ...     1  <p>I've read somewhere the human eye can't dis...
3  269         91.0  ...     4  <p>Yes, I thought about that, but I soon figur...
4  307         49.0  ...    28  <p><a href="http://www.codeproject.com/Article...

[5 rows x 6 columns]

In [7]:
tags_df.head()

Id             Tag
0  80            flex
1  80  actionscript-3
2  80             air
3  90             svn
4  90     tortoisesvn

In [8]:
users_df.head()

Unnamed: 0  id display_name  ... up_votes down_votes  views
0           0  26        Shawn  ...      565         67   3547
1           1  26        Shawn  ...      565         67   3547
2           2  26        Shawn  ...      565         67   3547
3           3  26        Shawn  ...      565         67   3547
4           4  26        Shawn  ...      565         67   3547

[5 rows x 9 columns]

# Data Cleaning

In [9]:
questions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1264216 entries, 0 to 1264215
Data columns (total 7 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   Id            1264216 non-null  int64  
 1   OwnerUserId   1249762 non-null  float64
 2   CreationDate  1264216 non-null  object 
 3   ClosedDate    55959 non-null    object 
 4   Score         1264216 non-null  int64  
 5   Title         1264216 non-null  object 
 6   Body          1264216 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 67.5+ MB


In [10]:
questions_df.isnull().sum() / len(questions_df) * 100

Id               0.000000
OwnerUserId      1.143317
CreationDate     0.000000
ClosedDate      95.573620
Score            0.000000
Title            0.000000
Body             0.000000
dtype: float64

1. As you can see above almost : **95.57 Percent** of rows are blank in cloumn **ClosedDate**

*   ***Case 1***: Those questions are still not closed.
*   ***Case 2***: They have not been updated in the csv.

We can take a Judgement and drop the column.

2. We are also not intrested, currently in dealing with Title and Body, Out of 
Scope for this EDA , so we can drop them as well. 








In [11]:
questions_df_mod = questions_df.drop(columns=['ClosedDate','Title','Body'])

In [12]:
questions_df_mod.head()

Id  OwnerUserId          CreationDate  Score
0   80         26.0  2008-08-01T13:57:07Z     26
1   90         58.0  2008-08-01T14:41:24Z    144
2  120         83.0  2008-08-01T15:50:08Z     21
3  180    2089740.0  2008-08-01T18:42:19Z     53
4  260         91.0  2008-08-01T23:22:08Z     49

In [13]:
questions_df_mod.isnull().sum() / len(questions_df_mod) * 100

Id              0.000000
OwnerUserId     1.143317
CreationDate    0.000000
Score           0.000000
dtype: float64

In [14]:
questions_df_mod [ questions_df_mod['OwnerUserId'].isnull() ]  # The Null Values in OwnerUserId can be someone who has not yet registered the acount in StackOverflow, So as of now we need to treat it as missing.

Id  OwnerUserId          CreationDate  Score
47           4170          NaN  2008-08-07T00:25:03Z      9
130         12330          NaN  2008-08-15T15:06:31Z      6
158         15190          NaN  2008-08-18T20:32:49Z      2
198         18250          NaN  2008-08-20T15:40:53Z      0
300         29100          NaN  2008-08-26T22:28:22Z      5
...           ...          ...                   ...    ...
1261359  40076260          NaN  2016-10-16T22:38:59Z      1
1261501  40079260          NaN  2016-10-10T15:10:51Z      0
1263664  40130890          NaN  2016-10-19T12:02:24Z     -4
1263908  40136590          NaN  2016-10-19T13:24:10Z      0
1264010  40138620          NaN  2016-10-19T18:06:37Z      0

[14454 rows x 4 columns]

In [15]:
answers_df.head()

Id  OwnerUserId  ... Score                                               Body
0   92         61.0  ...    13  <p><a href="http://svnbook.red-bean.com/">Vers...
1  124         26.0  ...    12  <p>I wound up using this. It is a kind of a ha...
2  199         50.0  ...     1  <p>I've read somewhere the human eye can't dis...
3  269         91.0  ...     4  <p>Yes, I thought about that, but I soon figur...
4  307         49.0  ...    28  <p><a href="http://www.codeproject.com/Article...

[5 rows x 6 columns]

In [16]:
answers_df.isnull().sum() #We can treat the OwnerUserID null value as missing, because they are 

Id                  0
OwnerUserId     13200
CreationDate        0
ParentId            0
Score               0
Body                0
dtype: int64

In [17]:
answers_df_mod = answers_df.drop(columns='Body') # Dropping Body as it is consuming memory and currently not required in the EDA.

In [18]:
answers_df_mod.head()

Id  OwnerUserId          CreationDate  ParentId  Score
0   92         61.0  2008-08-01T14:45:37Z        90     13
1  124         26.0  2008-08-01T16:09:47Z        80     12
2  199         50.0  2008-08-01T19:36:46Z       180      1
3  269         91.0  2008-08-01T23:49:57Z       260      4
4  307         49.0  2008-08-02T01:49:46Z       260     28

In [19]:
tags_df.head()

Id             Tag
0  80            flex
1  80  actionscript-3
2  80             air
3  90             svn
4  90     tortoisesvn

In [20]:
tags_df.isnull().sum() / len(tags_df) * 100

Id     0.000000
Tag    0.029672
dtype: float64

In [21]:
tags_df.isnull().sum()

Id        0
Tag    1113
dtype: int64

In [22]:
tags_df [ tags_df['Tag'].isnull() ] 

Id  Tag
947         32280  NaN
4755       140030  NaN
6892       191640  NaN
11107      283190  NaN
11570      291340  NaN
...           ...  ...
3722203  39916800  NaN
3726905  39953580  NaN
3729127  39970580  NaN
3731869  39992410  NaN
3733313  40002640  NaN

[1113 rows x 2 columns]

In [23]:
# Dropping unnamed tags, as they are very minimal and not needed for our analysis
tags_df = tags_df [ ~ (tags_df['Tag'].isnull()) ] 

In [24]:
tags_df.isnull().sum()

Id     0
Tag    0
dtype: int64

In [25]:
users_df.head()

Unnamed: 0  id display_name  ... up_votes down_votes  views
0           0  26        Shawn  ...      565         67   3547
1           1  26        Shawn  ...      565         67   3547
2           2  26        Shawn  ...      565         67   3547
3           3  26        Shawn  ...      565         67   3547
4           4  26        Shawn  ...      565         67   3547

[5 rows x 9 columns]

In [26]:
users_df.columns

Index(['Unnamed: 0', 'id', 'display_name', 'creation_date', 'last_access_date',
       'reputation', 'up_votes', 'down_votes', 'views'],
      dtype='object')

In [27]:
users_df.drop(columns='Unnamed: 0',inplace=True)

In [28]:
users_df.head()

id display_name               creation_date  ... up_votes  down_votes  views
0  26        Shawn  2008-08-01 12:18:14.52 UTC  ...      565          67   3547
1  26        Shawn  2008-08-01 12:18:14.52 UTC  ...      565          67   3547
2  26        Shawn  2008-08-01 12:18:14.52 UTC  ...      565          67   3547
3  26        Shawn  2008-08-01 12:18:14.52 UTC  ...      565          67   3547
4  26        Shawn  2008-08-01 12:18:14.52 UTC  ...      565          67   3547

[5 rows x 8 columns]

In [54]:
users_df.isnull().sum()

id                    0
display_name        136
creation_date         0
last_access_date      0
reputation            0
up_votes              0
down_votes            0
views                 0
dtype: int64

In [58]:
# We actualy dont need display_name in our analysis. So we can drop them

users_df.drop(columns='display_name',inplace=True)

In [59]:
users_df.isnull().sum()

id                  0
creation_date       0
last_access_date    0
reputation          0
up_votes            0
down_votes          0
views               0
dtype: int64

In [72]:
users_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1990480 entries, 0 to 1990479
Data columns (total 7 columns):
 #   Column            Dtype 
---  ------            ----- 
 0   id                int64 
 1   creation_date     object
 2   last_access_date  object
 3   reputation        int64 
 4   up_votes          int64 
 5   down_votes        int64 
 6   views             int64 
dtypes: int64(5), object(2)
memory usage: 106.3+ MB


In [60]:
users_df.head()

id               creation_date  ... down_votes  views
0  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
1  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
2  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
3  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
4  26  2008-08-01 12:18:14.52 UTC  ...         67   3547

[5 rows x 7 columns]

In [61]:
# We need to merge questions df and tags df using the Id column : 
questions_df_mod.head()

Id  OwnerUserId          CreationDate  Score
0   80         26.0  2008-08-01T13:57:07Z     26
1   90         58.0  2008-08-01T14:41:24Z    144
2  120         83.0  2008-08-01T15:50:08Z     21
3  180    2089740.0  2008-08-01T18:42:19Z     53
4  260         91.0  2008-08-01T23:22:08Z     49

In [62]:
tags_df.head()

Id             Tag
0  80            flex
1  80  actionscript-3
2  80             air
3  90             svn
4  90     tortoisesvn

In [64]:
df1 = questions_df_mod.merge(tags_df)

In [65]:
df1.head()

Id  OwnerUserId          CreationDate  Score             Tag
0  80         26.0  2008-08-01T13:57:07Z     26            flex
1  80         26.0  2008-08-01T13:57:07Z     26  actionscript-3
2  80         26.0  2008-08-01T13:57:07Z     26             air
3  90         58.0  2008-08-01T14:41:24Z    144             svn
4  90         58.0  2008-08-01T14:41:24Z    144     tortoisesvn

In [67]:
df1 [ df1['Id'] == 80 ] 

Id  OwnerUserId          CreationDate  Score             Tag
0  80         26.0  2008-08-01T13:57:07Z     26            flex
1  80         26.0  2008-08-01T13:57:07Z     26  actionscript-3
2  80         26.0  2008-08-01T13:57:07Z     26             air

In [68]:
tags_df [ tags_df['Id'] == 80 ] 

Id             Tag
0  80            flex
1  80  actionscript-3
2  80             air

In [69]:
df1.head()

Id  OwnerUserId          CreationDate  Score             Tag
0  80         26.0  2008-08-01T13:57:07Z     26            flex
1  80         26.0  2008-08-01T13:57:07Z     26  actionscript-3
2  80         26.0  2008-08-01T13:57:07Z     26             air
3  90         58.0  2008-08-01T14:41:24Z    144             svn
4  90         58.0  2008-08-01T14:41:24Z    144     tortoisesvn

In [ ]:
# Merging Answers and User with the OwnerUserId and id

In [70]:
users_df.head()

id               creation_date  ... down_votes  views
0  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
1  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
2  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
3  26  2008-08-01 12:18:14.52 UTC  ...         67   3547
4  26  2008-08-01 12:18:14.52 UTC  ...         67   3547

[5 rows x 7 columns]

In [71]:
# pd.merge(left=answers_df_mod,right=users_df,left_on='OwnerUserId',right_on='id',right_index=True,how='inner')

Id  OwnerUserId  ... down_votes   views
0              92         61.0  ...        136    6905
1504        33759         61.0  ...        136    6905
2259        48055         61.0  ...        136    6905
2419        51394         61.0  ...        136    6905
7338       142425         61.0  ...        136    6905
...           ...          ...  ...        ...     ...
2014438  40141871     169266.0  ...      19047  246000
2014439  40141916    1902910.0  ...          3      42
2014449  40142118    1007941.0  ...          0      41
2014462  40142266    1182620.0  ...         77    9537
2014475  40142501     323721.0  ...         93    1964

[1446276 rows x 12 columns]

In [74]:
answers_df_mod.columns

Index(['Id', 'OwnerUserId', 'CreationDate', 'ParentId', 'Score'], dtype='object')

In [76]:
answers_df_mod.rename(columns={'OwnerUserId':'id'}, inplace = True)

In [ ]:
pd.merge(answers_df_mod,users_df,on='id',how='outer')

In [ ]:
questions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1264216 entries, 0 to 1264215
Data columns (total 7 columns):
 #   Column        Non-Null Count    Dtype  
---  ------        --------------    -----  
 0   Id            1264216 non-null  int64  
 1   OwnerUserId   1249762 non-null  float64
 2   CreationDate  1264216 non-null  object 
 3   ClosedDate    55959 non-null    object 
 4   Score         1264216 non-null  int64  
 5   Title         1264216 non-null  object 
 6   Body          1264216 non-null  object 
dtypes: float64(1), int64(2), object(4)
memory usage: 67.5+ MB


In [ ]:
questions_df[ questions_df['Score'] == questions_df['Score'].max() ]

Id  ...                                               Body
4907  348170  ...  <p>I mistakenly added files using the command:...

[1 rows x 7 columns]


*   When a user raises a question with a **data science tag ('python', 'r' 'matlab', 'sas', 'excel', 'sql')**, Stack Overflow sends the notification to the relevant users.



In [ ]:
tags_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3750994 entries, 0 to 3750993
Data columns (total 2 columns):
 #   Column  Dtype 
---  ------  ----- 
 0   Id      int64 
 1   Tag     object
dtypes: int64(1), object(1)
memory usage: 57.2+ MB


In [ ]:
df_test = questions_df.merge(tags_df[['Id', 'Tag']])

In [ ]:
df_test.head()

Id  ...             Tag
0  80  ...            flex
1  80  ...  actionscript-3
2  80  ...             air
3  90  ...             svn
4  90  ...     tortoisesvn

[5 rows x 8 columns]

In [ ]:
df_test['Tag'].value_counts()

javascript           124155
java                 115212
c#                   101186
php                   98808
android               90659
                      ...  
ofxkinect                 1
r-extension               1
cocos2d-swift             1
alassetsgroup             1
named-constructor         1
Name: Tag, Length: 37034, dtype: int64

In [ ]:
df_test [ ( (df_test['Tag'] == 'python') | (df_test['Tag'] == 'r') | (df_test['Tag'] == 'matlab') | (df_test['Tag'] == 'sas') | (df_test['Tag'] == 'excel') | (df_test['Tag'] == 'sql')) ] 

Id  ...     Tag
7             120  ...     sql
67           1970  ...     sql
72           2120  ...     sql
99           2840  ...     sql
130          4110  ...     sql
...           ...  ...     ...
3750911  40142550  ...     sql
3750921  40142600  ...  python
3750940  40142840  ...  python
3750945  40142860  ...   excel
3750981  40143190  ...  python

[133660 rows x 8 columns]

In [ ]:
users_df.columns

Index(['Unnamed: 0', 'id', 'display_name', 'creation_date', 'last_access_date',
       'reputation', 'up_votes', 'down_votes', 'views'],
      dtype='object')

In [ ]:
users_df_mod =users_df.rename(columns = {'id':'Id'})
print(users_df_mod.columns)

Index(['Unnamed: 0', 'Id', 'display_name', 'creation_date', 'last_access_date',
       'reputation', 'up_votes', 'down_votes', 'views'],
      dtype='object')


In [ ]:
df3 = df_test.merge(users_df_mod[['Id','display_name']])

In [ ]:
df3.head()

Id  OwnerUserId  ...             Tag display_name
0   80         26.0  ...            flex        alumb
1   80         26.0  ...  actionscript-3        alumb
2   80         26.0  ...             air        alumb
3  120         83.0  ...             sql        hitec
4  120         83.0  ...             sql        hitec

[5 rows x 9 columns]

In [ ]:
df3 [ ( (df3['Tag'] == 'python') | (df3['Tag'] == 'r') | (df3['Tag'] == 'matlab') | (df3['Tag'] == 'sas') | (df3['Tag'] == 'excel') | (df3['Tag'] == 'sql')) ] 

Id  OwnerUserId  ...     Tag          display_name
3           120         83.0  ...     sql                 hitec
4           120         83.0  ...     sql                 hitec
400        1970        116.0  ...     sql          Tim Saunders
401        1970        116.0  ...     sql          Tim Saunders
402        1970        116.0  ...     sql          Tim Saunders
...         ...          ...  ...     ...                   ...
117046  6888430     411023.0  ...  python                Max G.
117114  6926590     769464.0  ...  python         Narinder Guru
117147  6945850     873448.0  ...     sql  Karthikeyan Ramasamy
117157  6949640     879615.0  ...  python               Moennig
117163  7016530     888350.0  ...  python                 Arpit

[2615 rows x 9 columns]

In [ ]:
# pd.merge(df3, questions_df,on='OwnerUserId')

In [ ]:
answers_df.head()

Id  OwnerUserId  ... Score                                               Body
0   92         61.0  ...    13  <p><a href="http://svnbook.red-bean.com/">Vers...
1  124         26.0  ...    12  <p>I wound up using this. It is a kind of a ha...
2  199         50.0  ...     1  <p>I've read somewhere the human eye can't dis...
3  269         91.0  ...     4  <p>Yes, I thought about that, but I soon figur...
4  307         49.0  ...    28  <p><a href="http://www.codeproject.com/Article...

[5 rows x 6 columns]